cpp/src/arrow/util/byte_stream_split_internal.cc

```diff
@@ -0,0 +1,120 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/byte_stream_split_internal.h"
+#include "arrow/util/dispatch_internal.h"
+
+#include <array>
+
+namespace arrow::util::internal {
+
+using ::arrow::internal::DispatchLevel;
+using ::arrow::internal::DynamicDispatch;
+
+/************************
+ *  Decode dispatching  *
+ ************************/
+
+template <int kNumStreams>
+struct ByteStreamSplitDecodeDynamic {
+  using FunctionType = decltype(&ByteStreamSplitDecodeScalar<kNumStreams>);
+  using Implementation = std::pair<DispatchLevel, FunctionType>;
+
+  constexpr static auto implementations() {
+    return std::array{
+        // Limited interest for performance on builds without a minimum SIMD level
+        // So we put SSE4.2 and NEON as a base level.
```

Comment:
The comment is incorrect. It's not that there's "limited interest in performance", it's that we always expect SSE4.2 or NEON to be available on the respective architectures (x86, ARM).